### DISCUSSION TITLE
How to troubleshoot the "Timestep too small" error when using the optocoupler 4N25 in Qucs-S?

### DISCUSSION THREAD

**Discussion by deralbert** [ra3xdh/qucs_s#678] [Q&A] [answered]
I am attempting to simulate the feedback circuit of a simple power supply with the flyback topology. This circuit includes a TL431 with the voltage divider (resistors R1 and R2). The TL431 (marked as `IC1`) will conduct current if the voltage exceeds 12 volts.

I did the following circuit (picture 1). When the voltage `V1` on the right-hand side is 12 volts or below, the TL431 conducts no current, and there are no errors in the simulation.
<img width="640" alt="Screenshot_124" src="https://github.com/ra3xdh/qucs_s/assets/909718/29e8485e-2f62-475b-8e28-a3d33a2d6661">

However, when I set the voltage `V1 `above 12 volts, IC1 should conduct the current. If I attempt to simulate it, I encounter a simulation error (picture 2).
<img width="640" alt="Screenshot_125" src="https://github.com/ra3xdh/qucs_s/assets/909718/91ae982e-e198-4cd5-b56b-8d2139ac8dd2">

I am using a custom TL431 Spice model, which works fine when not using the optocoupler. I am using the standard optocoupler model 4N25 from the library provided with Qucs-S.

I have limited experience in simulating circuits and am uncertain about the cause of the error. I have already attempted to increase the simulation time step (to 1e-3 or 0.5), but it did not resolve the issue. Could someone please help with this error? What steps can I take to resolve it? Thank you in advance!

The whole error message is:
```
Ngspice started...
Using SPARSE 1.3 as Direct Linear Solver
Note: Starting dynamic gmin stepping
Warning: Dynamic gmin stepping failed
Note: Starting true gmin stepping
Warning: True gmin stepping failed
Note: Starting source stepping
Warning: source stepping failed
Note: Transient op started
Error: Transient op failed, timestep too small


Error: The operating point could not be simulated successfully.
    Any of the following steps may fail.!


Note: No compatibility mode selected!


Circuit: * qucs 24.2.0  c:/users/name/.qucs/test_prj/feedbacknetwork.sch

Reducing trtol to 1 for xspice 'A' devices
Doing analysis at TEMP = 27.000000 and TNOM = 27.000000


Transient solution failed -

Last Node Voltages
------------------

Node                                   Last Voltage        Previous Iter
----                                   ------------        -------------
_net0                                       2.58158              2.58158
_net1                                       2.58158              2.58158
_net2                                          12.5                 12.5
_net3                                          12.5                 12.5
_net5                                       1.13165              1.13165
_net4                                        1.7043               1.7043
optled_voltage                             0.572652             0.572652
_net7                                       1.86322              7.11322 *
_net6                                      0.686805              5.93681 *
xic1.6                                     0.418173             0.418173
xic1.7                                    -0.981827            -0.981827
xic1.4                                        2.495                2.495
xic1.5                                     -61.4742             -61.4742
_net8                                     -0.336778              4.91322 *
redled_voltage                              1.17642              1.17642
xic2.x1.12                                   1.7043               1.7043
xic2.x1.10                              0.000268528          0.000268528
xic2.x1.11                              6.03009e-07          6.03009e-07
xic2.x1.6                               3.08803e-15                 5.25 *
h.xic2.x1.h1#branch                    -5.95389e-07         -5.95389e-07
v.xic2.x1.vm#branch                       0.0080639            0.0080639
eredled_voltage#branch                 -1.17642e-08         -1.17642e-08
eoptled_voltage#branch                 -5.72652e-09         -5.72652e-09
v2#branch                                 0.0023229            0.0023229
v.xic1.v1#branch                         0.00863656           0.00863656
v1#branch                               -0.00889181          -0.00889181
voptled_current#branch                   0.00863656           0.00863656
vtlrefcurrent#branch                    2.15132e-06          2.15132e-06
a$poly$e.xic1.e1#branch_1_0                 302.462              302.462


doAnalyses: TRAN:  Timestep too small; initial timepoint: trouble with node "_net7"


tran simulation(s) aborted
Warning from checkvalid: vector optled_voltage is not available or has zero length.

No. of Data Rows : 0
Reset re-loads circuit * qucs 24.2.0  c:/users/name/.qucs/test_prj/feedbacknetwork.sch

Circuit: * qucs 24.2.0  c:/users/name/.qucs/test_prj/feedbacknetwork.sch

ngspice-42 done
Simulation finished
```

**Comment by ra3xdh** [comment]
The convergence errors (timestep too small and similar) are generally not related to Qucs-S GUI itself and very hard to debug. There exists no common route. The reasons may be very different. For example wrong model parameters, missing parasitic RC, missing input/output resistance, wrong TRAN simulation parameters. In the presented circuit you have direct connection of the voltage source, LED, and BJT part of the optcoupler. This may cause infinite current and convergence error. Also 2s is may be very long simulation time for Ngspice. You may try to reduce timestep. I guess a simple DC bias simulation would be sufficient for your circuit because you don't have pulsed or AC sources. Also keep in mind that Qucs-S contains TL431 model shipped with application. Use the search function in the libraries. 

If nothing helps, this error should be reported to Ngspice team providing minimal netlist that fails to converge.

**Reply by deralbert** [reply]
Yes, there is a built-in TL431. However, I needed to use a custom model.

**Comment by ra3xdh** [comment, marked as answer]
Also V2-LED-IC1 loop has no connection to ground. This may be trouble for Ngspice, because it can consider it as the floating node.

**Reply by deralbert** [reply]
Indeed, this was the main error. After I added the ground, everything worked as it should.

Last question: I now have two different grounds in my circuit. Are they also treated as separate grounds internally/programmatically? In other words, something like GND1 and GND2? Could there be a potential difference between these two grounds?

<img width="640" alt="Screenshot_127" src="https://github.com/ra3xdh/qucs_s/assets/909718/6526f235-632a-4dc8-aee5-f0b6bb85ec72">

**Comment by ra3xdh** [comment]
> I now have two different grounds in my circuit.

SPICE doesn't support different ground. The GND in SPICE is some node which is considered as 0V. Both grounds on the schematic are connected to 0V.

**Comment by deralbert** [comment]
What would be the right place to ask questions like the above, since they are not related to Qucs-S?

I wanted to simulate a simple switched-mode power supply (flyback topology) in which the circuit with TL431 and the optocoupler was partly reused. Unfortunately, I now encounter the same error: "Time step is too small". I am sure that I am again missing something obvious and that an experienced eye will quickly identify it.

**Comment by ra3xdh** [comment]
Ask this question in this thread first. Then if nothing help go to Ngspice forum: https://sourceforge.net/p/ngspice/discussion/120973/ The switching power supply is not a simple circuit to simulate. You may require a ferrite core model that is not implemented for Ngspice. Also PWM controller model may be required.

**Comment by deralbert** [comment]
Okay, here is the circuit. The problem is that the simulation runs for a while and then ends with an error "Timestep too small". The time spent in simulation is enough to plot part of the output voltage (Uout). The entire .sch code is provided at the end of this message. The entire error message is also at the end of this message.

Picture "Flyback-ish circuit in Qucs-S"
<img width="640" alt="Screenshot_129" src="https://github.com/ra3xdh/qucs_s/assets/909718/ee931ab7-b861-4c1a-9b73-b2efc4bff4c2">

I attempted to simulate this circuit generated by PI Expert (Power Integrations). However, I simplified the circuit. For example, I didn't include the snubber circuit and several other components in the simulation model. I also wasn't particularly meticulous in calculating the values of the components for the simulation. My initial goal was to observe behavior that at least moves in the right direction. In the next step, I thought, I would tweak the component values to achieve the best performance.

Picture "SMPS (Flyback) generated by PI Expert"
<img width="640" alt="Screenshot_130" src="https://github.com/ra3xdh/qucs_s/assets/909718/fcd80d63-b987-488d-9466-9be965b37775">
I'll refrain from providing further details regarding the circuit for now, but I'm willing to provide them if requested.

**Schematic file Flyback.sch**: 
[Flyback.zip](https://github.com/ra3xdh/qucs_s/files/14922646/Flyback.zip)

**TL431 model**: [user_lib.zip](https://github.com/ra3xdh/qucs_s/files/14922617/user_lib.zip)

**Error message:**
```
Ngspice started...
Using SPARSE 1.3 as Direct Linear Solver
Note: Starting dynamic gmin stepping
Note: Dynamic gmin stepping completed

Note: No compatibility mode selected!

Circuit: * qucs 24.2.0  c:/users/Albert/.qucs/test_prj/flyback.sch

Reducing trtol to 1 for xspice 'A' devices
Doing analysis at TEMP = 27.000000 and TNOM = 27.000000


Initial Transient Solution
--------------------------

Node                                   Voltage
----                                   -------
_net2                              1.18698e-05
_net3                              5.74734e-05
uout                               5.74734e-05
_net4                              5.74734e-05
_net5                              5.88916e-05
xic1.x1.12                         5.88916e-05
_net6                              6.06645e-05
xic1.x1.10                        -2.91223e-18
xic1.x1.11                        -2.91223e-18
_net0                                        0
xic1.x1.6                         -2.12411e-08
xic2.6                                 1772.84
xic2.7                                 1771.44
xic2.4                                   2.495
xic2.5                                 1771.44
_net7                                        0
_net8                                      230
_net9                                      230
xt1._net5                                    0
xt1._net95                                 230
_net1                                      230
xt1._net4                         -4.03897e-28
xt1._net11                         2.53006e-12
xt1._net76                         2.28003e-12
xt1._net86                                 230
xt1._net8                          2.93007e-12
h.xic1.x1.h1#branch                          0
v.xic1.x1.vm#branch               -8.74545e-17
ltr1_l2#branch                     5.74896e-17
ltr1_l1#branch                     1.22018e-09
l.xt1.llg#branch                   4.60011e-10
l.xt1.lld#branch                  -1.22018e-09
l.xt1.lls#branch                    7.6001e-10
l1#branch                          5.99256e-10
euout#branch                      -5.74734e-13
v3#branch                         -1.22018e-09
v.xic2.v1#branch                   -1.7717e-09
v2#branch                          4.60011e-10
a$poly$e.xic2.e1#branch_1_0       -3.54339e-09

doAnalyses: TRAN:  Timestep too small; time = 5.08921e-06, timestep = 2.5e-18: trouble with node "xt1._net5"


tran simulation(s) aborted
ASCII raw file "spice4qucs.tr1.plot"
Reset re-loads circuit * qucs 24.2.0  c:/users/Albert/.qucs/test_prj/flyback.sch

Circuit: * qucs 24.2.0  c:/users/Albert/.qucs/test_prj/flyback.sch

ngspice-42 done
Simulation finished
```

**Comment by ra3xdh** [comment]
Try putting resistance between the MOSFET gate and ground. The closed BJT part of the optocoupler may have near infinite resistance. Also try the series resistance for PWM source, because the gate is capacitive load. Keep in mind that ideal transformer doesn't simulate the core saturation effect.

**Comment by ra3xdh** [comment]
The replacing of the MOSFET device also may help.

**Comment by deralbert** [comment]
Unfotunately, it didn't help, the same error occurs. I added R7 and R9 and replaced the T2 by the model IRF730.

> Try putting resistance between the MOSFET gate and ground.

I am not sure if I understood you correctly. There is no direct connection between the MOSFET's (T2) gate and ground.

<img width="640" alt="Screenshot_132" src="https://github.com/ra3xdh/qucs_s/assets/909718/36c42acf-93e0-4a6a-8007-f77087b03435">

**Comment by tomhajjar** [comment]
@deralbert 

If your going to ask the ngspice guys for help you need to provide a netlist, libraries, .spiceinit and a screenshot of the schematic. The netlist should also have plot statements added. You can look at the ngspice examples for how to format plot statements.

I would first "debug" the netlist using ngspice directly. If using Windows you can use the Ngspice GUI for MS Windows.
https://ngspice.sourceforge.io/experimental/ngspice_start.7z

PS

There are many TL431 Spice models available. Yours is probably the more "bare bone" one I have seen. This is OK as long as it works. I have tested a number of them that don't work properly especially in ngspice.

Also your schematic looks suspect...

**Reply by deralbert** [reply]
Thanks, I will try to debug the netlist.

I took the TL431 model from here: [https://www.ti.com/product/TL431#design-tools-simulation](https://www.ti.com/product/TL431#design-tools-simulation).

**Comment by ra3xdh** [comment]
@deralbert You may try to remove the feedback part and check if the convergence error still present. Also try to add snubber circuit. MOSFET may go into breakdown that also may cause convergence issues.

**Comment by tomhajjar** [comment]
@deralbert 

I would work on the schematic first before worrying about the netlist. The original circuit uses a switching IC. Your circuit doesn't come close to reflecting the ICs function. The values for R7/R9 are totally unrealistic. Transformer inductances don't seem correct. Primary inductance is much smaller that output inductance yet it's a step-down/buck configuration.

You should your attach project schematics, models/libraries/subcircuits minus data or PDFs. You didn't send the TL431 subcircuit schematic.

**Reply by deralbert** [reply]
I'll try to do it all step by step. Thank you.

**Reply by tomhajjar** [reply]
I kludged it to get it partially working. The schematic has multiple problems as stated above.

One would NEVER design a flyback like this. Please look at design examples from the IC and transformer manufacturers.

![2024-04-10_152653](https://github.com/ra3xdh/qucs_s/assets/102838604/cbb9e13b-dcc6-4d64-b6c8-c1b27b83c99e)

[Flyback_prj.zip](https://github.com/ra3xdh/qucs_s/files/14936857/Flyback_prj.zip)

**Reply by deralbert** [reply]
@tomhajjar

Your response is very valuable to me. I will now take a closer look and compare the circuits from the simulation and from PI Expert. Thank you very much.

I'm really bad at simulation. I can handle simple circuits, but when it comes to more complex ones, I struggle. In particular in this case, I simply attempted a "low-effort solution" hoping that either something halfway correct would come out or an error that I would understand. Also messed with inductance values. I think I just experimented around and forgot that I set the inductance value for the secondary part too high.

I'm not a specialist in the field of power electronics; I learn it for fun in my free time, mostly on my own. I discovered power electronics a few years ago, and it was love at first sight. I find that very interesting.

When it comes to practical application, I'm slightly better off. A similar power supply to the one I tried to simulate does what it's supposed to do. Here connected to a LED matrix from a light bulb, twelve LEDs I guess.
<img width="320" alt="Screenshot_132" src="https://github.com/ra3xdh/qucs_s/assets/909718/20576385-4e91-4133-bf3a-09b0cdd8ced2">
<img width="320" alt="Screenshot_132" src="https://github.com/ra3xdh/qucs_s/assets/909718/cdfd56f2-4100-4752-b201-18906464883f">
<img width="320" alt="Screenshot_132" src="https://github.com/ra3xdh/qucs_s/assets/909718/1f63fd83-739f-483f-b751-3d058a01d8be">
<img width="320" alt="Screenshot_132" src="https://github.com/ra3xdh/qucs_s/assets/909718/fe2f585f-2640-4be6-a227-aa72a5ffeb98">
